app/modules/mdns.c

```diff
@@ -12,125 +12,53 @@
-// passes are kept in step.
 //
-static wrapper_t *process_args(lua_State *L, phase_t phase, size_t *sizep)
+// mdns.register(hostname [, { attributes} ])
+//
+static int mdns_register(lua_State *L)
 {
-  wrapper_t *result = NULL;
-  char *p = NULL;
-  bool has_desc = !lua_isnumber(L, 3);
-  int arg_no = 1;
-
-  if (phase == PHASE_COPY_DATA) {
-    result = (wrapper_t *) c_zalloc(sizeof(wrapper_t) + *sizep);
-    if (!result) {
-      return NULL;
-    }
-    p = &result->data;
-  }
-
-  if (phase == PHASE_CALCULATE_LENGTH) {
-    luaL_checktype(L, arg_no, LUA_TSTRING);
-    *sizep += c_strlen(luaL_checkstring(L, arg_no++)) + 1;
-    luaL_checktype(L, arg_no, LUA_TSTRING);
-    *sizep += c_strlen(luaL_checkstring(L, arg_no++)) + 1;
-    if (has_desc) {
-      luaL_checktype(L, arg_no, LUA_TSTRING);
-      *sizep += c_strlen(luaL_checkstring(L, arg_no++)) + 1;
-    }
-    (void) luaL_checkinteger(L, arg_no++);
-  } else {
-    c_strcpy(p, luaL_checkstring(L, arg_no++)); 
-    result->mdns_info.host_name = p;
-    result->mdns_info.host_desc = p;
-    p = advance_over_string(p);
-
-    if (has_desc) {
-      c_strcpy(p, luaL_checkstring(L, arg_no++));
-      result->mdns_info.host_desc = p;
-      p = advance_over_string(p);
-    }
+  struct nodemcu_mdns_info info;
 
-    c_strcpy(p, luaL_checkstring(L, arg_no++));
-    result->mdns_info.service_name = p;
-    p = advance_over_string(p);
-
-    result->mdns_info.server_port = luaL_checkinteger(L, arg_no++);
-  }
+  memset(&info, 0, sizeof(info));
+  
+  info.host_name = luaL_checkstring(L, 1);
 
-  if (lua_gettop(L) >= arg_no) {
-    luaL_checktype(L, arg_no, LUA_TTABLE);
+  if (lua_gettop(L) >= 2) {
+    luaL_checktype(L, 2, LUA_TTABLE);
     lua_pushnil(L); // first key
     int slot = 0;
-    while (lua_next(L, arg_no) != 0 && slot < sizeof(result->mdns_info.txt_data) / sizeof(result->mdns_info.txt_data[0])) {
-      if (phase == PHASE_CALCULATE_LENGTH) {
-        luaL_checktype(L, -2, LUA_TSTRING);
-        *sizep += c_strlen(luaL_checkstring(L, -2)) + 1;
-        *sizep += c_strlen(luaL_checkstring(L, -1)) + 1;
+    while (lua_next(L, 2) != 0 && slot < sizeof(info.txt_data) / sizeof(info.txt_data[0])) {
+      luaL_checktype(L, -2, LUA_TSTRING);
+      const char *key = luaL_checkstring(L, -2);
+
+      if (c_strcmp(key, "port") == 0) {
+	info.service_port = luaL_checknumber(L, -1);
+      } else if (c_strcmp(key, "service") == 0) {
+	info.service_name = luaL_checkstring(L, -1);
+      } else if (c_strcmp(key, "description") == 0) {
+	info.host_desc = luaL_checkstring(L, -1);
       } else {
-        // put in the key
-	c_strcpy(p, luaL_checkstring(L, -2));
-        result->mdns_info.txt_data[slot++] = p;
-	p = advance_over_string(p);
-
-	// now smash in the value
+	int len = c_strlen(key) + 1;
 	const char *value = luaL_checkstring(L, -1);
-	p[-1] = '=';
-	c_strcpy(p, value);
-	p = advance_over_string(p);
+	char *p = alloca(len + c_strlen(value) + 1);
```

Comment:
`alloca` allocates a variable string on the stack frame which goes out of scope on the function exit and the contents are invalid, and hence the  `info.txt_data[]` entry becomes invalid.